apollo-federation/src/query_plan/query_planner.rs

```diff
@@ -490,11 +500,179 @@ impl QueryPlanner {
+        supergraph_schema
+            .schema()
+            .root_operation(operation.root_kind.into())
+            .and_then(|name| supergraph_schema.get_type(name.clone()).ok())
+            .and_then(|ty| ty.try_into().ok())
+    } else {
+        None
+    };
+    // We have to serially compute a plan for each top-level selection.
+    let mut split_roots = split_top_level_fields(operation.selection_set.clone());
+    let mut digest = Vec::new();
+    let mut starting_fetch_id = 0;
+    let selection_set = split_roots
+        .next()
+        .ok_or_else(|| FederationError::internal("Empty top level fields"))?;
+    let BestQueryPlanInfo {
+        mut fetch_dependency_graph,
+        path_tree: mut prev_path,
+        ..
+    } = compute_root_parallel_best_plan(parameters, selection_set, has_defers)?;
+    let mut prev_subgraph = only_root_subgraph(&fetch_dependency_graph)?;
+    for selection_set in split_roots {
+        let BestQueryPlanInfo {
+            fetch_dependency_graph: new_dep_graph,
+            path_tree: new_path,
+            ..
+        } = compute_root_parallel_best_plan(parameters, selection_set, has_defers)?;
+        let new_subgraph = only_root_subgraph(&new_dep_graph)?;
+        if new_subgraph == prev_subgraph {
+            // The new operation (think 'mutation' operation) is on the same subgraph than the previous one, so we can concat them in a single fetch
+            // and rely on the subgraph to enforce seriability. Do note that we need to `concat()` and not `merge()` because if we have
+            // mutation Mut {
+            //    mut1 {...}
+            //    mut2 {...}
+            //    mut1 {...}
+            // }
+            // then we should _not_ merge the 2 `mut1` fields (contrarily to what happens on queried fields).
+
+            prev_path = OpPathTree::merge(&prev_path, &new_path);
+            fetch_dependency_graph = FetchDependencyGraph::new(
+                supergraph_schema.clone(),
+                federated_query_graph.clone(),
+                root_type.clone(),
+                starting_fetch_id,
+            );
+            compute_root_fetch_groups(
+                operation.root_kind,
+                &mut fetch_dependency_graph,
+                &prev_path,
+            )?;
+        } else {
+            // TODO(@TylerBloom): Verify that this is correct? It could be the case that we just
+            // need to grab `.starting_id_generation`.
+            starting_fetch_id = fetch_dependency_graph.next_fetch_id();
+            digest.push(std::mem::replace(
+                &mut fetch_dependency_graph,
+                new_dep_graph,
+            ));
+            prev_path = new_path;
+            prev_subgraph = new_subgraph;
+        }
+    }
+    digest.push(fetch_dependency_graph);
+    Ok(digest)
+}
+
+fn split_top_level_fields(
+    selection_set: SelectionSet,
+) -> Box<dyn 'static + Iterator<Item = SelectionSet>> {
```

Comment:
I would much rather return an `impl Iterator` here, but that would require writing a recursive iterator for this function. It would avoid a few unnecessary allocations and avoid dynamic dispatch, though.Still unsure which would be better.

Reply:
I guess it's a pick your poison kind of scenario - recursive iterator here might (don't quote me) right into lifetime shenanigans. On the other, it would be nice to get a better handle on clones, because this function does it *a lot*.  We can also postpone this decision to later (make a ticket so we don't forget), and just keep what you have here for a few weeks.

An unrelated question: have you thought about having this directly on the SelectionSet struct rather than a standalone function? Are there considerations that prevent us doing something like that?

Reply:
> recursive iterator here might (don't quote me) right into lifetime shenanigans

Sorry for *literally* quoting you, but if everything is owned you should be ok (though, this might cause other problems like moving values out of `Arc`s). Otherwise, lifetimes will need to be thought through. Even this solution ran into 

> On the other, it would be nice to get a better handle on clones, because this function does it a lot

Very much agreed. While most of the clones are relatively cheap arc clones, I would love to avoid them and the v-table overhead. I also think a non-adhoc solution would have significantly fewer allocations. All of that does sound like something that can wait.

> An unrelated question: have you thought about having this directly on the SelectionSet struct rather than a standalone function?

I did not. It probably should be.

Reply:
@lrlna , I'm going to leave this as is and make a `TODO` comment about what should change. Do you want me to open a ticket as well?

Reply:
yes, please open a ticket! I'll DM you on slack with details